MM-20321: added logic for removing orphaned installations

gsagula opened the pull request:
This PR changes the backend logic of the list command. Installations in the KV store that have no reference in the cloud server database will be removed and user will be notified via direct message. Users will only see an error message when attempt to list an installation in the KV store that was marked as deleted but has no reference in the cloud server database.

Fixes: https://mattermost.atlassian.net/browse/MM-20231

Signed-off-by: Gabriel Linden Sagula <gsagula@gmail.com>

gabrieljackson commented on server/command_list.go:
```diff
@@ -28,23 +28,46 @@ func (p *Plugin) runListCommand(args []string, extra *model.CommandArgs) (*model
+		if cloudInstall == nil {
+			// If it was never an installation in cloud server and it was never marked as deleted
+			// in KV store, we return en error.
+			if pluginInstall.DeleteAt == 0 {
```
Hmm, I am not sure that this is correct. The deleteAt timestamp here is a value remembered by the plugin database. I am not sure why we would want to check this older value.

gsagula commented:
@gabrieljackson I've modified the logic for finding and removing orphaned installations. PTAL

gabrieljackson commented on server/command_list.go:
```diff
@@ -28,23 +27,43 @@ func (p *Plugin) runListCommand(args []string, extra *model.CommandArgs) (*model
+	}
+
+	for _, cloudInstall := range cloudInstalls {
+		for j, pluginInstall := range pluginInstalls {
```
Hmm, I could see this nested loop leading to long-term performance hits. Here is what I am am thinking.

The `cloudInstall` slice will slowly get larger and larger as the user creates more cloud installations. I know that it is unlikely that this will become a real issue, but considering this slice will only grow (in that even deleted installations are always returned) I think we can try something else.

We could keep logic similar to the original code where it assumes the installation is not deleted and asks for it. If it is returned as expected, we proceed as normal. If it isn't returned correctly, then make a follow-up query to make sure it was deleted. If it was deleted according to the provisioner, we cleanup the plugin store.

What do you think of this?

gsagula replied:
> We could keep logic similar to the original code where it assumes the installation is not deleted and asks for it. If it is returned as expected, we proceed as normal. If it isn't returned correctly, then make a follow-up query to make sure it was deleted. If it was deleted according to the provisioner, we cleanup the plugin store.

I think we can do that for now but it still does solve the problem with a large data set.  Also, if that plan is to never remove records related to deleted instances, I think we should also consider to fetch this data in chunks instead of the whole thing.

gsagula commented:
@gabrieljackson Thanks for the review. I added a comment for the potential performance issue.

gabrieljackson commented:
Hey @gigawhitlocks, just giving you a little ping on this. Guessing it got lost in your todo list.

gigawhitlocks commented on server/command_list.go:
```diff
@@ -28,23 +27,47 @@ func (p *Plugin) runListCommand(args []string, extra *model.CommandArgs) (*model
+	for _, cloudInstall := range cloudInstalls {
+		for j, pluginInstall := range pluginInstalls {
+			if cloudInstall.ID == pluginInstall.ID {
+				if cloudInstall.DeleteAt > 0 || cloudInstall.State == cloud.ClusterInstallationStateCreationFailed {
```
Lines 49-68 in this PR have a few problems. The first is that `if` statements are nested unnecessarily, when instead `continue` can be used strategically to flatten the routine and make it easier to reason about. Having the "happy path" on the left, achieved by checking error states first and then returning (or in this case `continue`ing) is also [more idiomatic Go](https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88) so I hope that's a style choice we can adhere to. I personally find code where the happy path is down the left easier to reason about.

Lines 58 through 61 also just don't work, see [this Playground link](https://play.golang.org/p/XGTzLuYXIUO). The logical error here is that the last element is always lost.

I've gone ahead and fixed these issues for you, because I was uncertain about my own concerns and wanted to verify before leaving this review. If we don't have a unit test for this method, we should probably write one before merging this to catch issues like the one on lines 58-61. 
```go
	for _, cloudInstall := range cloudInstalls {
		for j, pluginInstall := range pluginInstalls {
			if cloudInstall.ID != pluginInstall.ID {
				continue
			}

			if cloudInstall.DeleteAt > 0 || cloudInstall.State == cloud.ClusterInstallationStateCreationFailed {
				err = p.deleteInstallation(pluginInstall.ID)
				if err != nil {
					p.API.LogError(err.Error(), pluginInstall.ID)
					continue
				}
				// Notify the user that installation was removed.
				p.PostBotDM(userID, fmt.Sprintf("Cloud installation ID %s has been removed from your Mattermost app.", pluginInstall.ID))
				// Remove key from the plugin installations.
				pluginInstalls = append(pluginInstalls[:j], pluginInstalls[j+1:]...)
			} else {
				pluginInstall.Installation = *cloudInstall
				pluginInstall.License = "hidden"
			}

			break
		}
	}
```
I have only tested the line with append(), which you can see in the Playground link. The rest of the `if` statement flattening that I've done here should be tested.

gsagula replied:
Good catch, thanks @gigawhitlocks! I think I know what happened, but it definitely needs more testing. I have to go back to some old changes but I'm pretty sure that I commit some old code. I think that L59 shouldn't be there. If removed, the last element takes the place of the one being deleted. The whole idea is to avoid reallocations by using append function since I didn't know the since of these arrays. I will look at the rest of your comments and add more tests later, but it should work if L59 (which shouldn't be there at first place) is deleted.

gigawhitlocks replied:
Ah, I thought using `append` with a slice was smarter than it is, but it looks like you are right and `append` does copy the elements around. Your optimization, once the logic is corrected, will be more performant for sure, and I'm more than happy to see that checked in. I'd suggest creating a helper function for those 4 lines though, because then it should be trivial and worthwhile to write a unit test for the "remove element" function, since it is a little tricky.

gigawhitlocks replied:
Could you address some of the style issues in this block that I brought up, besides the bug fixed in commit 02c0087? This block of code is difficult to understand and I provided some refactored code to make it simpler. 

Regarding the bug fixed in 02c0087, I am concerned that changing the length of a list that Go is currently iterating over with `for ... range` will cause an index out of bounds exception. Is there a test for this code block? Considering its complexity and that we have found a bug in it already, I think this block of code needs to be refactored and the surrounding function should be exercised and protected with a test before it is merged.

See also this Playground link to see the behavior of removing an element from a list while iterating over it with `range`:
https://play.golang.org/p/wJrY8BLbtrl

gsagula replied:
Sorry, that's a problem caused by some immutability nature of `range`. Here is the fixed version:
https://play.golang.org/p/NxRHcVT3P8N 

I'm looking at the other parts of the comment.

gsagula commented:
@gabrieljackson I'm still writing unit tests here.

gabrieljackson commented:
I see that you started using Go Mocks for more unit tests. This makes sense for things like the docker and provisioner calls, but we were using the standard Mattermost plugin API mocks elsewhere already. Are those unable to test what you are trying to do?

gsagula commented:
@gabrieljackson I didn't see that, I will change it. Mattermost plugin API mocks are probably sufficient for what I'm testing. Thanks!

gsagula commented:
Actually, there is a number of problems of using the Mattermost mocked api for testing. It doesn't failed the test when method is called incorrectly or never called. It also has has some memory management issues since it's very difficult to reset the mocked objects. No wondering that the server project has so many flaky tests. I will use it because plugin is a mattermost-server related project, otherwise I would avoid it.

gsagula commented:
@gigawhitlocks Thank you for the review. I added the tests that you requested and fixed the bug, but I don't want to change too much the logic in code. I think this was more like a temporary fix than a permanent solution anyway (at least I hope). Feel free to take another pass!
